server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -8,32 +8,194 @@
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        WorkloadManagementSettings workloadManagementSettings
+    ) {
+
+        this(
+            taskCancellationService,
+            clusterService,
+            threadPool,
+            workloadManagementSettings,
+            new NodeDuressTrackers(
+                Map.of(
+                    ResourceType.CPU,
+                    new NodeDuressTracker(
+                        () -> workloadManagementSettings.getNodeLevelCpuCancellationThreshold() < ProcessProbe.getInstance()
+                            .getProcessCpuPercent() / 100.0,
+                        workloadManagementSettings::getDuressStreak
+                    ),
+                    ResourceType.MEMORY,
+                    new NodeDuressTracker(
+                        () -> workloadManagementSettings.getNodeLevelMemoryCancellationThreshold() <= JvmStats.jvmStats()
+                            .getMem()
+                            .getHeapUsedPercent() / 100.0,
+                        workloadManagementSettings::getDuressStreak
+                    )
+                )
+            ),
+            new HashMap<>(),
+            new HashSet<>(),
+            new HashSet<>()
+        );
     }
 
-    public QueryGroupService(Map<String, QueryGroupState> queryGroupStateMap) {
-        this.queryGroupStateMap = queryGroupStateMap;
+    public QueryGroupService(
+        QueryGroupTaskCancellationService taskCancellationService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        WorkloadManagementSettings workloadManagementSettings,
+        NodeDuressTrackers nodeDuressTrackers,
+        Map<String, QueryGroupState> stateMap,
+        Set<QueryGroup> activeQueryGroups,
+        Set<QueryGroup> deletedQueryGroups
+    ) {
+        this.taskCancellationService = taskCancellationService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.nodeDuressTrackers = nodeDuressTrackers;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        activeQueryGroups.forEach(queryGroup -> stateMap.putIfAbsent(queryGroup.get_id(), new QueryGroupState()));
+        this.queryGroupStateMap = stateMap;
+        this.queryGroupStateMap.put(QueryGroupTask.DEFAULT_QUERY_GROUP_ID_SUPPLIER.get(), new QueryGroupState());
+        taskCancellationService.setQueryGroupStateMapAccessor(this::getQueryGroupState);
+        clusterService.addStateApplier(this);
+    }
+
+    /**
+     * run at regular interval
+     */
+    protected void doRun() {
+        if (workloadManagementSettings.getWlmMode() == WlmMode.DISABLED) {
+            return;
+        }
+        taskCancellationService.refreshQueryGroups(activeQueryGroups, deletedQueryGroups);
+        taskCancellationService.cancelTasks(nodeDuressTrackers::isNodeInDuress);
```

Comment:
There is implicit assumption here between the order of `refreshQueryGroups`, `cancelTasks` and state maintained within `QueryGroupTaskCancellationService`? That looks slightly weird to me. I am wondering if can simplify this by having one method call.

Reply:
Makes sense, will make them the method param